lib/process_pilot/context.rb

```diff
@@ -35,10 +34,42 @@ def initialize(sources: nil, moddles: nil, services: {}, listeners: [], utils: {
       end
     end
 
+    def start(process_id: nil, start_event_id: nil, variables: {})
+      process = process_id ? process_by_id(process_id) : default_process
+      raise ExecutionError.new(process_id ? "Process #{process_id} not found." : "No default process found.") if process.blank?
+      execution = Execution.start(context: self, process: process, start_event_id: start_event_id, variables: variables)
+      executions << execution
+      execution
+    end
+
+    def start_with_message(message_name:, variables: {})
+      [].tap do |executions|
+        processes.map do |process|
+          process.start_events.map do |start_event|
+            start_event.message_event_definitions.map do |message_event_definition|
+              if message_name == message_event_definition.message_name
+                Execution.start(context: self, process: process, variables: variables, start_event_id: start_event.id).tap { |execution| executions.push execution }
+              end
+            end
+          end
+        end
+      end
+    end
+
+    def restore(execution_state)
+      Execution.deserialize(execution_state, context: self).tap do |execution|
+        executions << execution
+      end
+    end
+
     def notify_listener(event)
       listeners.each do |listener|
         listener[event[:event]].call(event) if listener[event[:event]]
-      end
+      end if listeners.present?
```

Comment:
Listeners is always an array.

```suggestion
      end
```